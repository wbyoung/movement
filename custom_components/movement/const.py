"""Constants for the Movement integration."""

import datetime as dt
from typing import Final

DOMAIN: Final = "movement"

ATTR_DISTANCE: Final = "distance_traveled"
ATTR_DISTANCE_ADJUSTMENTS: Final = "distance_adjustments"
ATTR_DISTANCE_WALKING: Final = "distance_walking"
ATTR_DISTANCE_BIKING: Final = "distance_biking"
ATTR_DISTANCE_DRIVING: Final = "distance_driving"
ATTR_SPEED: Final = "speed"
ATTR_MODE_OF_TRANSIT: Final = "mode_of_transit"
ATTR_GPS_ACCURACY: Final = "gps_accuracy"
ATTR_UPDATE_COUNT: Final = "distance_updates"

CONF_DEPENDENT_ENTITIES: Final = "dependent_entities"
CONF_TRACKED_ENTITY: Final = "tracked_entity"
CONF_TRIP_ADDITION: Final = "trip_addition"
CONF_MULTIPLIERS: Final = "multipliers"
CONF_NEIGHBORHOOD: Final = "neighborhood"
CONF_LOCAL: Final = "local"
CONF_HIGHWAY: Final = "highway"

UPDATES_STALLED_DELTA: Final = dt.timedelta(minutes=20)
HISTORY_EXPIRATION_DELTA: Final = dt.timedelta(hours=1)
SPEED_USABLE_DELTA: Final = dt.timedelta(seconds=45)
DEBOUNCE_UPDATES_DELTA: Final = dt.timedelta(seconds=5)

MAX_RESTORE_HISTORY: Final = 25
MAX_RESTORE_TRANSITION: Final = 25
