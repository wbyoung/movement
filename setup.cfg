[coverage:run]
source =
  custom_components

[coverage:report]
exclude_also =
    raise NotImplemented\(\)
    if __name__ == ['"]__main__["']:
show_missing = true

[tool:pytest]
testpaths = tests
norecursedirs = .git .mypy_cache
addopts =
    -p syrupy
    --strict
    --cov=custom_components
    --cov-branch
    --cov-report=term
    --cov-report=html

asyncio_mode = auto
asyncio_default_fixture_loop_scope = function

[flake8]
# https://github.com/ambv/black#line-length
max-line-length = 88
# E501: line too long
# W503: Line break occurred before a binary operator
# E203: Whitespace before ':'
# D202 No blank lines allowed after function docstring
# W504 line break after binary operator
ignore =
    E501,
    W503,
    E203,
    D202,
    W504

[isort]
# https://github.com/timothycrosley/isort
# https://github.com/timothycrosley/isort/wiki/isort-Settings
# splits long import on multiple lines indented by 4 spaces
multi_line_output = 3
include_trailing_comma=True
force_grid_wrap=0
use_parentheses=True
line_length=88
indent = "    "
# will group `import x` and `from x import` of the same module.
force_sort_within_sections = true
sections = FUTURE,STDLIB,THIRDPARTY,FIRSTPARTY,LOCALFOLDER
default_section = THIRDPARTY
known_first_party = custom_components,tests
forced_separate = tests
combine_as_imports = true

[mypy]
python_version = 3.13
follow_imports = normal
local_partial_types = true
strict_equality = true
no_implicit_optional = true
warn_incomplete_stub = true
warn_redundant_casts = true
warn_unused_configs = true
warn_unused_ignores = true
enable_error_code = ignore-without-code, redundant-self, truthy-iterable
disable_error_code = annotation-unchecked, import-not-found, import-untyped
extra_checks = false
check_untyped_defs = true
warn_return_any = true
warn_unreachable = true
explicit_package_bases = true
