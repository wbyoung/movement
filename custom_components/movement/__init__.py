"""The Movement integration."""

from __future__ import annotations

import logging

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.event import (
    async_track_entity_registry_updated_event,
    async_track_state_change_event,
)
from homeassistant.helpers.typing import ConfigType

from .const import CONF_DEPENDENT_ENTITIES, CONF_TRACKED_ENTITY, DOMAIN
from .coordinator import MovementUpdateCoordinator
from .services import async_setup_services
from .types import MovementConfigEntry, MovementConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SENSOR]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(  # noqa: RUF029
    hass: HomeAssistant,
    config: ConfigType,  # noqa: ARG001
) -> bool:
    """Set up Movement services.

    Returns:
        If the setup was successful.
    """
    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: MovementConfigEntry) -> bool:
    """Set up Movement from a config entry.

    Returns:
        If the setup was successful.
    """
    _LOGGER.debug("setup %s with config:%s", entry.title, entry.data)

    coordinator = MovementUpdateCoordinator(hass, entry)

    entry.async_on_unload(
        async_track_state_change_event(
            hass,
            [entry.data[CONF_TRACKED_ENTITY]],
            coordinator.async_handle_entity_state_change,
        ),
    )

    entry.async_on_unload(
        async_track_entity_registry_updated_event(
            hass,
            [
                entry.data[CONF_TRACKED_ENTITY],
                *entry.data.get(CONF_DEPENDENT_ENTITIES, []),
            ],
            coordinator.async_handle_tracked_entity_change,
        ),
    )

    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = MovementConfigEntryRuntimeData(
        coordinator=coordinator,
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
    return True


async def async_unload_entry(hass: HomeAssistant, entry: MovementConfigEntry) -> bool:
    """Unload a config entry.

    Returns:
        If the unload was successful.
    """
    unload_ok: bool = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if unload_ok:
        coordinator = entry.runtime_data.coordinator
        coordinator.cancel_all_listeners()

    return unload_ok


async def _async_update_listener(
    hass: HomeAssistant,
    entry: MovementConfigEntry,
) -> None:
    """Handle options update."""
    await hass.config_entries.async_reload(entry.entry_id)
