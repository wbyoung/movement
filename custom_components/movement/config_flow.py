"""Config flow for Movement integration."""

from __future__ import annotations

from typing import Any, Final, cast

from homeassistant.components.device_tracker import DOMAIN as DEVICE_TRACKER_DOMAIN
from homeassistant.components.person import DOMAIN as PERSON_DOMAIN
from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN, SensorStateClass
from homeassistant.components.template import DOMAIN as TEMPLATE_DOMAIN
from homeassistant.config_entries import (
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
    OptionsFlow,
)
from homeassistant.const import UnitOfLength
from homeassistant.core import State, callback
from homeassistant.data_entry_flow import section
from homeassistant.helpers import entity_registry as er
from homeassistant.helpers.selector import (
    EntitySelector,
    EntitySelectorConfig,
    NumberSelector,
    NumberSelectorConfig,
    NumberSelectorMode,
)
from homeassistant.helpers.typing import VolDictType
from homeassistant.util import slugify
import voluptuous as vol

from .const import (
    CONF_DEPENDENT_ENTITIES,
    CONF_HIGHWAY,
    CONF_LOCAL,
    CONF_MULTIPLIERS,
    CONF_NEIGHBORHOOD,
    CONF_TRACKED_ENTITY,
    CONF_TRIP_ADDITION,
    DOMAIN,
)

SECTION_ADVANCED_OPTIONS: Final = "advanced_options"


def _base_schema(user_input: dict[str, Any]) -> VolDictType:
    return {
        vol.Required(
            CONF_TRACKED_ENTITY, default=user_input.get(CONF_TRACKED_ENTITY, [])
        ): EntitySelector(
            EntitySelectorConfig(domain=[DEVICE_TRACKER_DOMAIN, PERSON_DOMAIN]),
        ),
        vol.Required(
            CONF_TRIP_ADDITION,
            default=user_input.get(CONF_TRIP_ADDITION, 0),
        ): NumberSelector(
            NumberSelectorConfig(
                min=0, max=100, step=0.01, unit_of_measurement=UnitOfLength.KILOMETERS
            ),
        ),
        vol.Required(CONF_MULTIPLIERS): section(
            vol.Schema(
                {
                    vol.Required(
                        CONF_NEIGHBORHOOD,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(
                            CONF_NEIGHBORHOOD, 1
                        ),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                    vol.Required(
                        CONF_LOCAL,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(CONF_LOCAL, 1),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                    vol.Required(
                        CONF_HIGHWAY,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(
                            CONF_HIGHWAY, 1
                        ),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                }
            ),
            {"collapsed": True},
        ),
    }


class MovementConfigFlow(ConfigFlow, domain=DOMAIN):  # type: ignore[call-arg]
    """Handle a Movement config flow."""

    VERSION = 1

    def _user_form_schema(self) -> vol.Schema:
        return vol.Schema(_base_schema({}))

    @staticmethod
    @callback
    def async_get_options_flow(config_entry: ConfigEntry) -> OptionsFlow:
        """Get the options flow for this handler."""
        return MovementOptionsFlow()

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user."""
        if user_input is not None:
            self._async_abort_entries_match(user_input)

            title = cast(
                State, self.hass.states.get(user_input[CONF_TRACKED_ENTITY])
            ).name

            slugified_existing_entry_titles = [
                slugify(e.title) for e in self._async_current_entries()
            ]

            possible_title = title
            tries = 1
            while slugify(possible_title) in slugified_existing_entry_titles:
                tries += 1
                possible_title = f"{title} {tries}"

            return self.async_create_entry(title=possible_title, data=user_input)

        return self.async_show_form(
            step_id="user",
            data_schema=self._user_form_schema(),
        )


class MovementOptionsFlow(OptionsFlow):
    """Handle a option flow."""

    def _user_form_schema(self, user_input: dict[str, Any]) -> vol.Schema:
        entity_registry = er.async_get(self.hass)
        return vol.Schema(
            {
                **_base_schema(user_input),
                vol.Required(SECTION_ADVANCED_OPTIONS): section(
                    vol.Schema(
                        {
                            vol.Required(
                                CONF_DEPENDENT_ENTITIES,
                                default=user_input.get(CONF_DEPENDENT_ENTITIES, []),
                            ): EntitySelector(
                                EntitySelectorConfig(
                                    integration=TEMPLATE_DOMAIN,
                                    domain=[SENSOR_DOMAIN],
                                    include_entities=[
                                        entity_id
                                        for entity_id in entity_registry.entities.keys()
                                        if _is_total_increasing_and_kilometers(
                                            entity_registry, entity_id
                                        )
                                    ],
                                    multiple=True,
                                ),
                            )
                        }
                    ),
                    {"collapsed": True},
                ),
            }
        )

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle options flow."""
        if user_input is not None:
            advanced_options = user_input.pop(SECTION_ADVANCED_OPTIONS)
            self.hass.config_entries.async_update_entry(
                self.config_entry,
                data={**self.config_entry.data, **user_input, **advanced_options},
            )
            return self.async_create_entry(title=self.config_entry.title, data={})

        return self.async_show_form(
            step_id="init",
            data_schema=self._user_form_schema(dict(self.config_entry.data)),
        )


def _is_total_increasing_and_kilometers(
    entity_registry: er.EntityRegistry,
    entity_id: str,
) -> bool:
    """Determine if an entity_id represents an entity with a `state_class` of
    `total_increasing` and a `unit_of_measurement` of `km`."""
    return bool(
        (entry := entity_registry.async_get(entity_id))
        and (
            entry.capabilities
            and entry.capabilities.get("state_class")
            == SensorStateClass.TOTAL_INCREASING
            and entry.unit_of_measurement == UnitOfLength.KILOMETERS
        )
    )
