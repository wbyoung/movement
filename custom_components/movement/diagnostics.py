"""Diagnostics support for Movement."""

from collections.abc import Callable, Mapping
import random
from typing import Any, cast, overload

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ACCESS_TOKEN,
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
)
from homeassistant.core import HomeAssistant, callback

from .coordinator import MovementUpdateCoordinator

CONF_REFRESH_TOKEN = "refresh_token"  # noqa: S105
CONF_VIN = "vin"

TO_REDACT = {
    CONF_API_KEY,
    CONF_ACCESS_TOKEN,
    CONF_REFRESH_TOKEN,
    CONF_VIN,
}


async def async_get_config_entry_diagnostics(  # noqa: RUF029
    hass: HomeAssistant,  # noqa: ARG001
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: MovementUpdateCoordinator = entry.runtime_data.coordinator
    data: dict[str, Any] = {
        "entry": entry.as_dict(),
        "data": coordinator.data.as_dict(),
        "history": [item.as_dict() for item in coordinator.history.items],
        "transition": [item.as_dict() for item in coordinator.transition.items]
        if coordinator.transition.items is not None
        else None,
    }

    add_latitude = random.uniform(0, 180)  # noqa: S311
    add_longitude = random.uniform(0, 360)  # noqa: S311

    return map_data(
        async_redact_data(data, TO_REDACT),
        {
            CONF_LATITUDE: lambda x: mod_latitude(x + add_latitude),
            CONF_LONGITUDE: lambda x: mod_longitude(x + add_longitude),
        },
    )


@overload
def map_data(
    data: Mapping, mappers: Mapping[Any, Callable[[Any], Any]]
) -> dict: ...  # pragma: no cover


@overload
def map_data[T](data: T, mappers: Mapping[Any, Callable[[Any], Any]]) -> T: ...


@callback
def map_data[T](data: T, mappers: Mapping[Any, Callable[[Any], Any]]) -> T:
    """Map values in a dict.

    Returns:
        The data with values mapped based on the given mappers.
    """
    if not isinstance(data, (Mapping, list)):
        return data

    if isinstance(data, list):
        return cast("T", [map_data(val, mappers) for val in data])

    result = {**data}

    for key, value in result.items():
        if key in mappers:
            result[key] = mappers[key](result[key])
        elif isinstance(value, Mapping):
            result[key] = map_data(value, mappers)
        elif isinstance(value, list):
            result[key] = [map_data(item, mappers) for item in value]

    return cast("T", result)


def mod_latitude(value: float) -> float:
    return (90 + value) % 180 - 90


def mod_longitude(value: float) -> float:
    return (180 + value) % 360 - 180
