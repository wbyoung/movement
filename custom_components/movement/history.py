"""Helper to manage history."""

from dataclasses import replace
import datetime as dt
import logging
from typing import Callable

from homeassistant.core import State
from homeassistant.util import dt as dt_util

from .const import DEBOUNCE_UPDATES_DELTA, HISTORY_EXPIRATION_DELTA, SPEED_USABLE_DELTA
from .types import ABSENT_NONE, AbsentNone, HistoryEntry, Location, StateChangedData

HISTORY_ENTRIES_MAX = 1000


_LOGGER = logging.getLogger(__name__)
_TRACE = 5


class HistoryRegistry:
    """HistoryRegistry class."""

    def __init__(self):
        super().__init__()
        self._items: list[HistoryEntry] = []
        self._prior: list[HistoryEntry] | None = None

    @property
    def items(self) -> list[HistoryEntry]:
        return self._items

    @property
    def prior(self) -> list[HistoryEntry]:
        if self._prior is None:
            raise AttributeError(
                "prior items unavailable: registry was (re-)initialized and has had no new entries added"
            )

        return self._prior

    @property
    def current_entry(self) -> HistoryEntry:
        return self.items[0]

    @property
    def prior_entry(self) -> HistoryEntry:
        return self.prior[0]

    def reset(self, items: list[HistoryEntry]):
        self._items = items
        self._prior = None

    def add_entry_from_state_change(
        self,
        change: StateChangedData,
        unworkable: Callable[[str], None] = lambda reason: None,
    ):
        """
        Update the location history with a new entry at the head of the list.

        Args:
            change: The state changed data from which to derive the new entry.
            unworkable: A callable to invoke if an unworkable data set is
                generated.

        Filter out older and less accurate entries from the remainder of the list
        while keeping all items that are not yet within the usable threshold for
        speed calculations (and one more just beyond that threshold) (see
        consideration #2 in `MovementUpdateCoordinator._calculate_speed` for why
        keeping recent items is important).

        Also, constrain the history to a maximum size.
        """
        _LOGGER.log(_TRACE, "add_entry_from_state_change %s", change)

        # this generally won't happen frequently for users, but could happen if
        # a person entity removes all of their devices that they're using for
        # tracking. but there will be no prior location for many tests cases.
        fallback_entry: HistoryEntry = self._make_fallback_entry(change)
        unworkable_reason: str | None = None

        self._items = self._clean(self._items)
        self._prune_items(change.new_state.attributes.get("gps_accuracy", None))
        self._items.insert(0, self._make_entry(change.new_state))
        self._mark_debounced_items()
        self._items = self._items[:HISTORY_ENTRIES_MAX]  # constrain the history
        self._prior = self._clean(self._items[1:]) or ([fallback_entry])

        # flag to skip updates marked for debounce/inaccuracy
        if self._items[0].debounce:
            unworkable_reason = "first history item marked for debounce"
        elif self._items[0].ignore == "inaccurate":
            unworkable_reason = "poor gps_accuracy"
        elif (ignore := self._prior[0].ignore) and self._prior[0] == fallback_entry:
            unworkable_reason = f"poor details ({ignore}) in `from_state`"

        if unworkable_reason:
            unworkable(unworkable_reason)

    def _make_entry(self, state: State, at: dt.datetime | None = None) -> HistoryEntry:
        attrs = state.attributes
        gps_accuracy = attrs.get("gps_accuracy")
        inaccurate = _default_negative(gps_accuracy) > 1000
        ignore: str | AbsentNone = ABSENT_NONE
        latitude = attrs.get("latitude")
        longitude = attrs.get("longitude")
        location: Location | AbsentNone = Location(
            latitude=latitude,
            longitude=longitude,
        )

        if inaccurate:
            ignore = "inaccurate"
            location = ABSENT_NONE

        if not latitude or not longitude:
            ignore = "no_location"
            location = ABSENT_NONE

        return HistoryEntry(
            at=at or dt_util.utcnow(),
            location=location,
            accuracy=_default_no_accuracy(gps_accuracy),
            ignore=ignore,
        )

    def _make_fallback_entry(self, change: StateChangedData) -> HistoryEntry:
        """Make the fallback history entry."""
        return self._make_entry(change.old_state, at=change.old_state.last_changed)

    def _prune_items(self, gps_accuracy: float | None):
        """Prune items for usability.

        Items are always kept if:

        - Not enough time has passed (SPEED_USABLE_DELTA) to make them usable
          yet for speed calculations

        Otherwise, they're kept if:

        - Have better accuracy than more recent entries and haven't expired
        - They're the first non-expired item that can be used for speed
          calculations even if they have worse accuracy than the given
          `gps_accuracy`.
        """
        current_time = dt_util.utcnow()
        usable_before = current_time - SPEED_USABLE_DELTA
        expire_time = current_time - HISTORY_EXPIRATION_DELTA

        have_one_usable = False
        accuracy_threshold = _default_negative(gps_accuracy)

        result = []

        for index, entry in enumerate(self._items):
            usable = entry.at <= usable_before
            waiting_to_be_usable = not usable
            expired = entry.at <= expire_time
            is_accurate = entry.accuracy < accuracy_threshold
            keep = waiting_to_be_usable or (is_accurate and not expired)

            # always keep the first one that's not expired that can be used for speed
            # calculations
            if usable and not expired and not have_one_usable:
                keep = True
                have_one_usable = True

            if keep:
                result.append(entry)
                accuracy_threshold = min(accuracy_threshold, entry.accuracy)

            _LOGGER.debug(
                f"\n"
                f"  [{index}]\n"
                f"    usable: {usable}\n"
                f"    waiting_to_be_usable: {waiting_to_be_usable}\n"
                f"    expired: {expired}\n"
                f"    is_accurate: {is_accurate}\n"
                f"    have_one_usable: {have_one_usable}\n"
                f"    keep: {keep}"
            )

        self._items = result

    def _mark_debounced_items(self):
        """Mark items that should be debounced.

        Entries that arrive with a very short timeframe of each other are marked
        (and later cleaned out of the working set of items) to work around
        home-assistant/core#126972.
        """
        result: list[HistoryEntry] = []
        prev_entry = None
        for index, entry in enumerate(reversed(self._items)):
            if prev_entry is not None:
                delta = entry.at - prev_entry.at
                _LOGGER.debug("[%s] delta for debounce: %s", index, delta)
                if delta < DEBOUNCE_UPDATES_DELTA:
                    entry = replace(entry, debounce=True)
            result.insert(0, entry)
            prev_entry = entry
        self._items = result

    @staticmethod
    def _clean(items: list[HistoryEntry]) -> list[HistoryEntry]:
        """Filter out entries that may be preserved for context.

        Some entries are left in the registry between updates just for context,
        but no longer provide any useful information. They generally do not
        provide useful information for calculations, but can be helpful in
        debugging.

        Both `ignore` and `debounce` should only be set to truthy values to
        make these filters easier to write.
        """
        return [item for item in items if not (item.ignore or item.debounce)]


def _default_negative(value: float | None) -> float:
    return -1 if value is None else value


def _default_no_accuracy(value: float | None) -> float:
    return HistoryEntry.NO_ACCURACY if value is None else value
