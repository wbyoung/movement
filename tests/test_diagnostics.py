"""Test Movement diagnostics."""

import datetime as dt
from typing import Any
from unittest.mock import patch

from freezegun import freeze_time
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component
import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_fire_time_changed,
)
from pytest_homeassistant_custom_component.components.diagnostics import (
    get_diagnostics_for_config_entry,
)
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator
from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props
from syrupy.matchers import path_type

from custom_components.movement.diagnostics import map_data

from . import MOCK_UTC_NOW, setup_integration


@pytest.mark.parametrize(
    "changes",
    [
        [],
        [(46.723, {"latitude": 35.052, "longitude": 137.145})],
        [
            (46.723, {"latitude": 35.052, "longitude": 137.145}),
            (53.821, {"latitude": 35.05, "longitude": 137.147}),
        ],
    ],
    ids=[
        "no_activity",
        "no_movement",
        "some_movement",
    ],
)
async def test_entry_diagnostics(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    hass_client: ClientSessionGenerator,
    snapshot: SnapshotAssertion,
    changes: list[tuple[float, dict[str, Any] | None]],
) -> None:
    """Test config entry diagnostics."""
    await async_setup_component(hass, "homeassistant", {})

    initial_attrs = {
        "friendly_name": "Akio Toyoda",
        "latitude": 35.054,
        "longitude": 137.143,
        "gps_accuracy": 5,
    }

    hass.states.async_set(
        "person.akio_toyoda",
        "not_home",
        initial_attrs,
    )

    with freeze_time(MOCK_UTC_NOW) as freezer:
        await setup_integration(hass, mock_config_entry)

        for tick, change_attrs in changes:
            freezer.tick(dt.timedelta(seconds=tick))
            async_fire_time_changed(hass)

            if change_attrs:
                hass.states.async_set(
                    "person.akio_toyoda",
                    "not_home",
                    initial_attrs | change_attrs,
                )

            await hass.async_block_till_done()

    with patch(
        "random.uniform",
        return_value=True,
    ) as mock_random:
        mock_random.side_effect = [-30, 22]

        assert await get_diagnostics_for_config_entry(
            hass, hass_client, mock_config_entry
        ) == snapshot(
            exclude=props("created_at", "modified_at"),
            matcher=_round_floats_matcher,
        )


@pytest.mark.parametrize(
    ("data", "mappers", "expected"),
    [
        ([], {}, []),
        ({}, {}, {}),
        ("simple-string", {}, "simple-string"),
        (
            {"map_me": 1, "keep_me": 1},
            {"map_me": lambda x: f"changed-{x}"},
            {"map_me": "changed-1", "keep_me": 1},
        ),
    ],
    ids=[
        "empty_list",
        "empty_object",
        "string",
        "simple-conversion",
    ],
)
def test_map_data(
    data: Any,
    mappers: dict[str, Any],
    expected: Any,
) -> None:
    assert map_data(data, mappers) == expected


_round_floats_matcher = path_type(
    types=(float,),
    replacer=lambda data, _: round(data, 5),
)
