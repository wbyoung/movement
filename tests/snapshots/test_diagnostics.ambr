# serializer version: 1
# name: test_entry_diagnostics[no_activity]
  dict({
    'data': dict({
      'adjustments': 0,
      'change_count': -1,
      'distance': 0,
      'ignore_count': 0,
      'mode_of_transit': None,
      'speed': None,
    }),
    'entry': dict({
      'data': dict({
        'multipliers': dict({
          'highway': 1,
          'local': 1,
          'neighborhood': 1,
        }),
        'tracked_entity': 'person.akio_toyoda',
        'trip_addition': 0,
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'movement',
      'entry_id': 'mock-entry-id',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'Mock Title',
      'unique_id': None,
      'version': 1,
    }),
    'history': list([
    ]),
    'transition': None,
  })
# ---
# name: test_entry_diagnostics[no_movement]
  dict({
    'data': dict({
      'adjustments': 0,
      'change_count': 1,
      'distance': 0.2874,
      'ignore_count': 0,
      'mode_of_transit': None,
      'speed': None,
    }),
    'entry': dict({
      'data': dict({
        'multipliers': dict({
          'highway': 1,
          'local': 1,
          'neighborhood': 1,
        }),
        'tracked_entity': 'person.akio_toyoda',
        'trip_addition': 0,
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'movement',
      'entry_id': 'mock-entry-id',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'Mock Title',
      'unique_id': None,
      'version': 1,
    }),
    'history': list([
      dict({
        'accuracy': 5,
        'at': '2025-05-20 10:52:18.726245+00:00',
        'location': dict({
          'latitude': 5.052,
          'longitude': 159.145,
        }),
      }),
    ]),
    'transition': list([
      dict({
        'distance': 0.2874,
      }),
    ]),
  })
# ---
# name: test_entry_diagnostics[some_movement]
  dict({
    'data': dict({
      'adjustments': 0.0,
      'change_count': 2,
      'distance': 0.57481,
      'ignore_count': 0,
      'mode_of_transit': 'driving',
      'speed': 19.22412,
    }),
    'entry': dict({
      'data': dict({
        'multipliers': dict({
          'highway': 1,
          'local': 1,
          'neighborhood': 1,
        }),
        'tracked_entity': 'person.akio_toyoda',
        'trip_addition': 0,
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'movement',
      'entry_id': 'mock-entry-id',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': 'Mock Title',
      'unique_id': None,
      'version': 1,
    }),
    'history': list([
      dict({
        'accuracy': 5,
        'at': '2025-05-20 10:53:12.547245+00:00',
        'location': dict({
          'latitude': 5.05,
          'longitude': 159.147,
        }),
      }),
      dict({
        'accuracy': 5,
        'at': '2025-05-20 10:52:18.726245+00:00',
        'location': dict({
          'latitude': 5.052,
          'longitude': 159.145,
        }),
      }),
    ]),
    'transition': list([
      dict({
        'adjustments': 0.0,
        'distance': 0.2874,
      }),
    ]),
  })
# ---
